File: src/web/movimentation.spa/src/app/@core/models/movimentation.ts
export class Movimentation {
  id: string;
  customerName: string;
  cpf: string;
  movimentationDate: string;
  tributeCode: string;
  tributeDescription: string;
  tributeAliquot: number;
  movimentationGain: number;
  movimentationLoss: number;

  constructor(id: string = "", customerName: string, cpf: string, movimentationDate: string, tributeCode: string, tributeDescription: string, tributeAliquot: number, movimentationGain: number) {
    this.id = id;
    this.customerName = customerName;
    this.cpf = cpf;
    this.movimentationDate = movimentationDate;
    this.tributeCode = tributeCode;
    this.tributeDescription = tributeDescription;
    this.tributeAliquot = tributeAliquot;
    this.movimentationGain = movimentationGain;
  }

  isValid(): Array<Error> {
    var errors = new Array<Error>();

    if (!this.customerName)
      errors.push(new Error("O nome do cliente deve ser preenchido"));

    if (!this.cpf)
      errors.push(new Error("O cpf deve ser preenchido"));

    if (!this.movimentationDate)
      errors.push(new Error("A data da movimentação deve ser preenchido"));

    if (!this.isValidDate(this.movimentationDate))
      errors.push(new Error("A data da movimentação está em um formato inválido"));

    if (!this.tributeCode)
      errors.push(new Error("O código do tributo deve ser preenchido"));

    if (!this.tributeDescription)
      errors.push(new Error("A descrição do tributo deve ser preenchido"));

    if (!this.tributeAliquot)
      errors.push(new Error("A alíquota de tributo deve ser preenchida"));

    if (!this.movimentationGain)
      errors.push(new Error("O valor rendimento da movimentação deve ser preenchido"));

    return errors;
  }

  isValidDate(data: string): boolean {
    return new Date(data).toUTCString() !== 'Invalid Date';
  }
}

File: src/services/Movimentation/MIU.Movimentation.Domain/Entities/Movimentation.cs
using MIU.Core.Domain;
using MIU.Core.ValueObjects;
using System;

namespace MIU.Movimentations.Domain.Entities
{
    public class Movimentation : Entity, IAggregateRoot
    {
        public Movimentation(string tributeCode, 
                             string customerName,
                             string cpf, 
                             DateTime movimentationDate, 
                             string tributeDescription, 
                             int tributeAliquot, 
                             int movimentationGain)
        {
            TributeCode = tributeCode;
            CustomerName = customerName;
            Cpf = new CPF(cpf);
            MovimentationDate = movimentationDate;
            TributeDescription = tributeDescription;
            TributeAliquot = tributeAliquot;
            MovimentationGain = movimentationGain;
            GenerateMovimentationLoss();
        }

        protected Movimentation(){ }

        public string TributeCode { get; set; }
        public string CustomerName { get; private set; }
        public CPF Cpf { get; private set; }
        public DateTime MovimentationDate { get; private set; }
        public string TributeDescription { get; private set; }
        public int TributeAliquot { get; private set; }
        public int MovimentationGain { get; private set; }
        public int MovimentationLoss { get; private set; }

        public void UpdateMovimentation(string tributeCode,
                                        string customerName,
                                        string cpf,
                                        DateTime movimentationDate,
                                        string tributeDescription,
                                        int tributeAliquot,
                                        int movimentationGain)
        {
            TributeCode = tributeCode;
            CustomerName = customerName;
            Cpf = new CPF(cpf);
            MovimentationDate = movimentationDate;
            TributeDescription = tributeDescription;
            TributeAliquot = tributeAliquot;
            MovimentationGain = movimentationGain;
            GenerateMovimentationLoss();
        }

        public void GenerateMovimentationLoss()
        {
            MovimentationLoss = (int)Math.Round((double)(TributeAliquot * MovimentationGain) / 100);
            MovimentationGain = MovimentationGain - (int)Math.Round((double)(TributeAliquot * MovimentationGain) / 100);
        }
    }
}
